dynamic-check/src/objects/frame.rs

```diff
@@ -0,0 +1,1198 @@
+                    *element_type
+                } else {
+                    Type::Array(element_type, dimensions)
+                };
+
+                (type_, count)
+            }
+            SpreadOrExpression::Expression(expression) => (self.parse_expression(expression)?, 1),
+        })
+    }
+
+    ///
+    /// Returns the type of the accessed array element when called as an expression.
+    ///
+    fn parse_expression_array_access(
+        &mut self,
+        expression: &Expression,
+        r_or_e: &RangeOrExpression,
+        span: &Span,
+    ) -> Result<Type, FrameError> {
+        // Parse the array expression which could be a variable with type array.
+        let type_ = self.parse_expression(expression)?;
+
+        // Parse the array access.
+        self.parse_array_access(type_, r_or_e, span)
+    }
+
+    ///
+    /// Returns the type of the accessed array element.
+    ///
+    fn parse_array_access(&mut self, type_: Type, r_or_e: &RangeOrExpression, span: &Span) -> Result<Type, FrameError> {
+        // Check the type is an array.
+        let (element_type, _dimensions) = match type_ {
+            Type::Array(type_, dimensions) => (type_, dimensions),
+            type_ => return Err(FrameError::array_access(&type_, span)),
+        };
+
+        // Get the length of the array.
+        // let length = *dimensions.last().unwrap();
+
+        // Evaluate the range as an array type or the expression as the element type.
+        match r_or_e {
+            RangeOrExpression::Range(from, to) => {
+                if let Some(expression) = from {
+                    // Parse the expression type.
+                    let type_ = self.parse_expression(expression)?;
+
+                    self.assert_index(&type_, span);
+                }
+
+                if let Some(expression) = to {
+                    // Parse the expression type.
+                    let type_ = self.parse_expression(expression)?;
+
+                    self.assert_index(&type_, span);
+                }
+            }
+            RangeOrExpression::Expression(expression) => {
+                // Parse the expression type.
+                let type_ = self.parse_expression(expression)?;
+
+                // Assert the type is an index.
+                self.assert_index(&type_, span);
+            }
+        }
+
+        Ok(*element_type)
+    }
+
+    ///
+    /// Returns the Self type of the frame or an error if it does not exist.
+    ///
+    fn self_type_or_error(&self, span: &Span) -> Result<CircuitType, FrameError> {
+        self.self_type
+            .as_ref()
+            .map(|circuit_type| circuit_type.clone())
+            .ok_or_else(|| FrameError::circuit_self(span))
+    }
+
+    ///
+    /// Returns the type of inline circuit expression.
+    ///
+    fn parse_circuit(
+        &mut self,
+        identifier: &Identifier,
+        members: &Vec<CircuitVariableDefinition>,
```

Comment:
```suggestion
        members: impl ExactSizeIterator<Item = CircuitVariableDefinition>,
```
if possible; otherwise this should be `&[CircuitVariableDefinition]`